californium-integration-tests/src/test/java/org/eclipse/californium/integration/test/SecureTest.java

```diff
@@ -0,0 +1,118 @@
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Simon Bernard (Sierra Wireless) - initial implementation.
+ ******************************************************************************/
+package org.eclipse.californium.integration.test;
+
+import static org.eclipse.californium.core.test.MessageExchangeStoreTool.assertAllExchangesAreCompleted;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.californium.TestTools;
+import org.eclipse.californium.category.Medium;
+import org.eclipse.californium.core.CoapClient;
+import org.eclipse.californium.core.network.EndpointManager;
+import org.eclipse.californium.core.network.config.NetworkConfig;
+import org.eclipse.californium.core.network.config.NetworkConfig.Keys;
+import org.eclipse.californium.core.test.CountingHandler;
+import org.eclipse.californium.core.test.MessageExchangeStoreTool.CoapTestEndpoint;
+import org.eclipse.californium.elements.StrictDtlsEndpointContextMatcher;
+import org.eclipse.californium.integration.test.util.CoapsNetworkRule;
+import org.eclipse.californium.scandium.DTLSConnector;
+import org.eclipse.californium.scandium.config.DtlsConnectorConfig;
+import org.eclipse.californium.scandium.config.DtlsConnectorConfig.Builder;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+@Category(Medium.class)
+public class SecureTest {
+
+	@ClassRule
+	public static CoapsNetworkRule network = new CoapsNetworkRule(CoapsNetworkRule.Mode.DIRECT,
+			CoapsNetworkRule.Mode.NATIVE);
+
+	// CoAP config constants
+	private static final int TEST_EXCHANGE_LIFETIME = 247; // milliseconds
+	private static final int TEST_SWEEP_DEDUPLICATOR_INTERVAL = 100; // milliseconds
+
+	// DTLS config constants
+	private static final String PSK_IDENITITY = "client1";
+	private static final String PSK_KEY = "key1";
+	private static final int NB_RETRANSMISSION = 2;
+	private static final int RETRANSMISSION_TIMEOUT = 100; // milliseconds
+
+	private CoapTestEndpoint coapTestEndpoint;
+
+	@Before
+	public void startupServer() {
+		System.out.println(System.lineSeparator() + "Start " + getClass().getSimpleName());
+	}
+
+	@After
+	public void shutdownServer() {
+		System.out.println("End " + getClass().getSimpleName());
+	}
+
+	/**
+	 * Ensure there is no leak when we try to send a request to an absent peer
+	 */
+	@Test
+	public void testSecureGetHandshakeTimeout() throws Exception {
+		// Create an endpoint
+		createEndpoint();
+
+		// Send a request to an absent peer
+		CoapClient client = new CoapClient("coaps", InetAddress.getLoopbackAddress().getHostAddress(),
+				TestTools.findFreePort());
```

Comment:
Instead of introducing "findFreePort" (and additionally using "setReuseAddress(true)" there), I would prefer tot open the DatagramSocket(0) in the test start, use it's assigned port for the test, and shutdown the socket in the test shutdown. And please, no "setReuseAddress(true)".

Reply:
What is the problem with setReuseAddress(true) ? 
It seems you spend bad time because of it :)

Reply:
In rare case on unix, the client and server have been operating at the same port, and then failed with extremely irritating logs. And in other OSs it may even not be effective at all. So it adds a lot of undetermined behaviour. Yes, I guess it will work 100 times without an error, but as more as you repeat the test, you will see the reds :-).